### PR TITLE
Upload File to IPFS and Mint NFT to Opensea

### PR DESCRIPTION
**Pull request by rykci**
Include scripts to pack directory into .car or .zip

Two options for uploading file:
1. Upload file to own IPFS network and generate metadata. (set gateway variables in .env)
2. Upload file using MCP API and generate metadata. (set mcp variables in .env)

Both methods will mint NFT via DatabaseMinter smart contract (run deploy task beforehand)

### CODE REVIEW

File: tasks/upload.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    // the name of the .car file without the .car extension
+    const metadataName = fileName.split('.')[0]
+
+    // metadata file object for IPFS api
+    const fileMetadata = {
+      path: `${metadataName}.json`, // add JSON extension
+      content: JSON.stringify({
+        name: taskArgs.name || metadataName, // directory name
+        description: taskArgs.desc,
+        //image: `${process.env.READ_GATEWAY}QmQbF9mJEYUdLaWgw568abFiwvR1udQsfmuLyhejTiZ2DG`, // placeholder database icon for opensea
```
Comment (liushmh) [inline tasks/upload.js]: remove this

File: tasks/pack.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+const fs = require('fs').promises
+const { FsBlockStore } = require('ipfs-car/blockstore/fs')
+const { task } = require('hardhat/config')
+
+task('pack', 'converts file or directory to .car file')
+  .addParam('dir', 'the path to the directory')
+  .setAction(async (taskArgs) => {
+    await packToFs({
+      input: taskArgs.dir,
+      output: `${taskArgs.dir}/../${taskArgs.dir.split('/').pop()}.car`,
```
Comment (liushmh) [inline tasks/pack.js]: is that possible to create env config for output?

File: tasks/upload.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  .addOptionalParam('name', 'name of the NFT')
+  .addOptionalParam('desc', 'NFT description')
+  .setAction(async (taskArgs) => {
+    // create an instance of the HTTP API client
+    const ipfs = create(process.env.WRITE_GATEWAY)
+
+    let path = taskArgs.file
+
+    // fileName is the name of the .car file
+    const fileName = path.split('/').pop()
```
Comment (liushmh) [inline tasks/upload.js:18]: could you do some check before using fileName, such as undefined and length of filename

File: tasks/mcp.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  const MinterContract = await ethers.getContractFactory('DatabaseMinter')
+
+  //Get signer information
+  const [signer] = await ethers.getSigners()
+
+  //Mint NFT
+  console.log('Minting...')
+  const databaseMinter = await MinterContract.attach(contractAddr)
+  const tx = await databaseMinter.mintData(signer.address, metadataCid)
```
Comment (liushmh) [inline tasks/mcp.js]: could you add err handling for mint?

File: tasks/mint.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    const MinterContract = await ethers.getContractFactory('DatabaseMinter')
+
+    //Get signer information
+    const [signer] = await ethers.getSigners()
+
+    //Mint NFT
+    console.log('Minting...')
+    const databaseMinter = await MinterContract.attach(contractAddr)
+    const tx = await databaseMinter.mintData(signer.address, taskArgs.cid)
```
Comment (liushmh) [inline tasks/mint.js]: err handling

File: tasks/upload.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            display_type: 'number',
+            trait_type: 'Size',
+            value: uploadResult.size,
+          },
+        ],
+      }),
+    }
+
+    // upload to IPFS and return object with size and CID
+    console.log(await ipfs.add(fileMetadata))
```
Comment (liushmh) [inline tasks/upload.js]: possible err handling

File: tasks/mcp.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  })
+
+const mcpUpload = async (fileName, file, duration = 180) => {
+  const form = new FormData()
+  form.append('duration', duration)
+  form.append('file', file, fileName)
+
+  let result = { cid: '' }
+
+  await axios
```
Comment (liushmh) [inline tasks/mcp.js]: you can take a look on how to use async/await for axios
it's like you combine async/await with returning Promise

File: tasks/upload.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    const fileName = path.split('/').pop()
+
+    // file object for IPFS api
+    const file = {
+      path: fileName, //get path name
+      content: await fs.readFile(path), //.car file
+    }
+
+    // upload to IPFS and return object with size and CID
+    const uploadResult = await ipfs.add(file, { wrapWithDirectory: true })
```
Comment (liushmh) [inline tasks/upload.js]: possible err  handle of ipfs uploading

File: tasks/mcp.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      console.log(JSON.stringify(metadata))
+
+      // upload JSON to MCP
+      console.log('Uploading metadata to MCP...')
+      const metadataUpload = await mcpUpload(
+        metadata.name,
+        JSON.stringify(metadata),
+      )
+
+      await mint(metadataUpload.cid) // mint NFT!
```
Comment (liushmh) [inline tasks/mcp.js]: possible err handling

**Review by liushmh — COMMENTED** [review]
please add some err handling, and pay attention to  not mix aysnc/await and promise in js

File: tasks/upload.js (hunk 3 above)
Comment (rykci) [reply tasks/upload.js:18]: Do we have minimum length?

**Review by liushmh — APPROVED** [review]

File: tasks/upload.js (hunk 1 above)
Comment (rykci) [reply tasks/upload.js]: removed

**Pull request merged**